rust_icu_ubrk/src/lib.rs

```diff
@@ -0,0 +1,735 @@
+        &self,
+        type_: sys::ULocDataLocaleType,
+    ) -> Result<String, common::Error> {
+        let mut status = common::Error::OK_CODE;
+        let char_ptr = unsafe {
+            assert!(common::Error::is_ok(status));
+            versioned_function!(ubrk_getLocaleByType)(
+                self.rep.as_ptr(),
+                type_,
+                &mut status,
+            )
+        };
+        common::Error::ok_or_warning(status)?;
+        let c_str = unsafe { ffi::CStr::from_ptr(char_ptr) };
+        let s = c_str.to_str().map(|s| s.to_owned())?;
+        Ok(s)
+    }
+
+    /// Returns the status of the break rule that determined the most-recently
+    /// returned boundary. The default status for rules that do not explicitly
+    /// provide one is zero.
+    ///
+    /// See the [ICU user guide](https://unicode-org.github.io/icu/userguide/boundaryanalysis/break-rules.html)
+    /// for details on rule syntax and rule status values.
+    ///
+    /// Implements `ubrk_getRuleStatus`.
+    pub fn get_rule_status(&self) -> i32 {
+        unsafe { versioned_function!(ubrk_getRuleStatus)(self.rep.as_ptr()) }
+    }
+
+    /// Returns the statuses of the break rules that determined the most-recently
+    /// returned boundary. The default status for rules that do not explicitly
+    /// provide one is zero.
+    ///
+    /// See the [ICU user guide](https://unicode-org.github.io/icu/userguide/boundaryanalysis/break-rules.html)
+    /// for details on rule syntax and rule status values.
+    ///
+    /// Implements `ubrk_getRuleStatusVec`.
+    pub fn get_rule_status_vec(&self) -> Result<Vec<i32>, common::Error> {
+        let mut status = common::Error::OK_CODE;
+        // Preflight to determine buffer size.
+        let rules_len = unsafe {
+            assert!(common::Error::is_ok(status));
+            versioned_function!(ubrk_getRuleStatusVec)(
+                self.rep.as_ptr(),
+                0 as *mut i32,
+                0,
+                &mut status,
+            )
+        };
+        common::Error::ok_preflight(status)?;
+        let mut status = common::Error::OK_CODE;
+        let mut rules: Vec<i32> = vec![0; rules_len as usize];
+        unsafe {
+            assert!(common::Error::is_ok(status));
+            versioned_function!(ubrk_getRuleStatusVec)(
+                self.rep.as_ptr(),
+                rules.as_mut_ptr(),
+                rules_len,
+                &mut status,
+            );
+        }
+        common::Error::ok_or_warning(status)?;
+        Ok(rules)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::UBreakIterator;
+    use log::trace;
+    use rust_icu_sys::{UBreakIteratorType::*, ULocDataLocaleType::*};
+    use std::{convert::TryFrom, rc::Rc};
+
+    const TEXT: &str =
+        r#""It wasn't the wine," murmured Mr. Snodgrass. "It was the salmon.""#;
+
+    const WORD_BOUNDARIES: [i32; 30] = [
+        0, 1, 3, 4, 10, 11, 14, 15, 19, 20, 21, 22, 30, 31, 33, 34, 35, 44, 45,
```

Comment:
It is a bit hard to check that the result is correct when just numeric.  Could you give a test case that also gives an example of using the returned breaks to extract text fragments?  What would it take to convert the word boundaries into an iterator of `&str`for example, that shows how `TEXT` was split?

Reply:
Good idea. I think iterating over `&str` also aligns more closely with a typical use case for a break iterator. I will address this next.

Reply:
I tried to address this in cea1b59. Please let me know if it doesn't align with what you had in mind.